src/main/java/at/ac/tuwien/kr/alpha/common/FunctionTerm.java

```diff
@@ -1,67 +1,96 @@
 package at.ac.tuwien.kr.alpha.common;
 
-import java.util.Arrays;
-import java.util.HashMap;
-import java.util.LinkedList;
-import java.util.List;
+import at.ac.tuwien.kr.alpha.grounder.Substitution;
+import com.google.common.collect.Interner;
+import com.google.common.collect.Interners;
+
+import java.util.*;
 
 import static at.ac.tuwien.kr.alpha.Util.appendDelimited;
 
 /**
  * Copyright (c) 2016, the Alpha Team.
  */
 public class FunctionTerm extends Term {
-	private static HashMap<FunctionTerm, FunctionTerm> knownFunctionTerms = new HashMap<>();
+	private static final Interner<FunctionTerm> INTERNER = Interners.newStrongInterner();
+
+	private final Symbol symbol;
+	private final List<Term> terms;
+	private final boolean ground;
+
+	private FunctionTerm(Symbol symbol, List<Term> terms) {
+		if (symbol == null) {
+			throw new IllegalArgumentException();
+		}
+
+		this.symbol = symbol;
+		this.terms = Collections.unmodifiableList(terms);
+
+		boolean ground = true;
+		for (Term term : terms) {
+			if (!term.isGround()) {
+				ground = false;
+				break;
+			}
+		}
+		this.ground = ground;
+	}
 
-	public final TermSymbol functionSymbol;
-	public final List<Term> termList;
+	public static FunctionTerm getInstance(Symbol functionSymbol, List<Term> termList) {
+		return INTERNER.intern(new FunctionTerm(functionSymbol, termList));
+	}
+
+	public static FunctionTerm getInstance(String functionSymbol, List<Term> termList) {
+		return getInstance(Symbol.getInstance(functionSymbol, termList.size()), termList);
+	}
 
-	private FunctionTerm(TermSymbol functionSymbol, List<Term> termList) {
-		this.functionSymbol = functionSymbol;
-		this.termList = termList;
+	public static FunctionTerm getInstance(String functionSymbol, Term... terms) {
+		return getInstance(Symbol.getInstance(functionSymbol, terms.length), Arrays.asList(terms));
 	}
 
-	public static FunctionTerm getFunctionTerm(TermSymbol functionSymbol, List<Term> termList) {
-		FunctionTerm functionTerm = new FunctionTerm(functionSymbol, termList);
-		knownFunctionTerms.putIfAbsent(functionTerm, functionTerm);
-		return knownFunctionTerms.get(functionTerm);
+	public List<Term> getTerms() {
+		return terms;
 	}
 
-	public static FunctionTerm getFunctionTerm(String functionSymbol, List<Term> termList) {
-		return getFunctionTerm(TermSymbol.getInstance(functionSymbol), termList);
+	public Symbol getSymbol() {
+		return symbol;
 	}
 
-	public static FunctionTerm getFunctionTerm(String functionSymbol, Term... terms) {
-		return getFunctionTerm(TermSymbol.getInstance(functionSymbol), Arrays.asList(terms));
+	public boolean isSimilar(FunctionTerm other) {
+		return symbol.equals(other.symbol);
```

Comment:
I removed this in favor of an explicit check in `Substitution`: `a.getSymbol().equals(b.getSymbol())`. `isSimilar()` was only called once anyways.